### ISSUE TITLE
Redesign automl recipe

### ISSUE THREAD

**Issue by yushan111** [intel-analytics/analytics-zoo#3772]
The future usage of recipe might be
```python
# user customized model
def model_creator(config):
    model = tf.keras.models.Sequential([tf.keras.layers.Dense(config["hidden_size"])])
    model.compile()
    return model

# user customized search space corresponding to model hyper parameters
from zoo.orca.automl import hp
search_space = {
    # model parameters
    "hidden_size": hp.grid_search([32, 64, 128]),

    # optimization parameters
    "lr": hp.uniform(0.001, 0.01),
    "batch_size": hp.choice([32, 64, 128])}

# build auto estimator and fit
auto_est = AutoEstimator.from_keras(model_creator=model_creator,...)
auto_est.fit(data, 
             search_space=search_space,
             metric="mse",
             num_samples=1,
             early_stop={"max_iter": 5, "threshold": 0.01}, # same as stop in tune.run(),
             search_alg=,

            )

**Comment by yushan111** [comment]
Breakdown tasks

- [ ] wrap all tune random distributions API in `zoo.orca.automl.hp`, change corresponding examples/documents.
- [ ] move `recipe.runtime_params` out into `searcher.compile` and expose in AutoEstimator.fit()
- [ ] move feature selection part into zouwu predictor and remove `all_available_features`  in `Recipe.search_space()`
- [ ] support model selection (with `hp.choice`?)

**Comment by TheaperDeng** [comment]
Update 4/14/2021:
Breakdown tasks
- [ ] wrap all tune random distributions API in `zoo.orca.automl.hp`, change corresponding examples/documents. (#3776 Running Jenkins)
- [ ] move `recipe.runtime_params` out into `searcher.compile` and expose in `AutoEstimator.fit()` 

Currently we report three key in our ray tune trainable function: `training_iteration`, `reward_metric` and `checkpoint`. The first two can be seen as runtime stop recipe. `num_samples` I believe can be got from normal search_space with default value = 1. (Not sure)
BTW, the API we use to report `training_iteration`, `reward_metric` and `checkpoint` has deprecated. #3787

**Comment by yushan111** [comment]
> Currently we report three key in our ray tune trainable function: `training_iteration`, `reward_metric` and `checkpoint`. The first two can be seen as runtime stop recipe. `num_samples` can be exposed to AutoEstimator as well.

`report` is our internal implementation. We can focus on how we expect user use our `recipe`. We used to have `training_iteration`, `num_samples` and `reward_metric` in our constructor of `BaseRecipe`. You can find the redesigned usage in `auto_est.fit()` of the first comment. Maybe we should use user specified metric name instead of `reward_metric`. 

> BTW, the API we use to report `training_iteration`, `reward_metric` and `checkpoint` has deprecated. #3787
Sure, you can fix that.

**Comment by TheaperDeng** [comment]
We can change all the built-in recipe to dictionary (or maybe two; one for search space, one for early stop) after PR #3844 ?

**Comment by jason-dai** [comment]
> The future usage of recipe might be
> 
> ```python
> # user customized model
> def model_creator(config):
>     model = tf.keras.models.Sequential([tf.keras.layers.Dense(config["hidden_size"])])
>     model.compile()
>     return model
> 
> # user customized search space corresponding to model hyper parameters
> from zoo.orca.automl import hp
> search_space = {
>     # model parameters
>     "hidden_size": hp.grid_search([32, 64, 128]),
> 
>     # optimization parameters
>     "lr": hp.uniform(0.001, 0.01),
>     "batch_size": hp.choice([32, 64, 128])}
> 
> # build auto estimator and fit
> auto_est = AutoEstimator.from_keras(model_creator=model_creator,...)
> auto_est.fit(data, 
>              search_space=search_space,
>              metric="mse",
>              num_samples=1,
>              early_stop={"max_iter": 5, "threshold": 0.01}, # same as stop in tune.run(),
>              search_alg=,
> 
>             )
> ```

for `fit`:
```python
auto_est.fit(data, 
             search_space=search_space,
             metric="mse",
             num_samples=1,
             early_stop={"max_iter": 5, "threshold": 0.01}, # same as stop in tune.run(),
             search_alg=,
            )
```
- Do we really need a dictionary for `early_stop`? If these two parameters are mandatory, we can simply make them arguments for `fit`
- It is not very clearly what `num_samples` and `max_iter" mean; change to something more explicit (e.g., `n_sampling`, etc.)

**Comment by yushan111** [comment]
@jason-dai 
For fit, shall we change to 
```python
auto_est.fit(data, 
             search_space=search_space,
             metric="mse",
             metric_threshold=None,
             n_sampling=1,
             max_epochs=1,
             search_alg=None,
            )
